### ISSUE TITLE
Agent gets stuck in updating throughout while upgrade with invalid agent binary.

### ISSUE THREAD

**Issue by amolnater-qasource** [elastic/elastic-agent#2317]
**Kibana version:**  8.7 BC3 kibana cloud environment

**Host OS and Browser version:** All, All

**Build details:**
```
VERSION: 8.7 BC3
BUILD: 60804
COMMIT: abbdcf43f4e89f6fc085360252514aed7c032b4b
```

**Preconditions:**
1. 8.7 BC3 kibana cloud environment should be available.
2. Lower version 8.6.2 agent should be installed.

**Steps to reproduce:**
1. Navigate to Fleet>Settings and add invalid agent binary: `https://test-artifacts.elastic.co/downloads/`
2. Navigate to Fleet>Agents tab.
3. Trigger agent upgrade for lower version agent.
4. Observe agent never gets failed and remains stuck in `Updating` throughout.

**Expected Result:**
Agent should fail while upgrade with invalid agent binary and should reset agent for  re-upgrade after sometime, say 10 minutes.

**Logs:**
[elastic-agent-diagnostics-2023-02-24T03-37-26Z-00.zip](https://github.com/elastic/elastic-agent/files/10820956/elastic-agent-diagnostics-2023-02-24T03-37-26Z-00.zip)

**Note:**
-  We have observed that agent remained in `Updating` state for over 12 hours.

**Screenshots:**
![image](https://user-images.githubusercontent.com/77374876/221101343-5d6fc93c-d702-4a03-8c05-ee433e1d8ef0.png)
![image](https://user-images.githubusercontent.com/77374876/221101359-b8629ace-7fe2-4a0c-baee-4b4b65e8dff7.png)
![9](https://user-images.githubusercontent.com/77374876/221101395-a9b5371e-03c1-4c69-b960-420dccf35137.JPG)
![image](https://user-images.githubusercontent.com/77374876/221101468-1bd6473a-8099-4ce9-870b-d8d5bbc63b02.png)

**Comment by amolnater-qasource** [comment]
@manishgupta-qasource Please review.

**Comment by manishgupta-qasource** [comment]
Secondary review for this ticket is Done

**Comment by cmacknz** [comment]
The upgrade on the agent side starts and immediately fails with:

```json
{"log.level":"info","@timestamp":"2023-02-23T19:12:04.841Z","log.origin":{"file.name":"upgrade/upgrade.go","file.line":116},"message":"Upgrading agent","log":{"source":"elastic-agent"},"version":"8.7.0","source_uri":"","ecs.version":"1.6.0"}
{"log.level":"warn","@timestamp":"2023-02-23T19:12:04.849Z","log.logger":"transport","log.origin":{"file.name":"transport/tcp.go","file.line":52},"message":"DNS lookup failure \"test-artifacts.elastic.co\": lookup test-artifacts.elastic.co: no such host","ecs.version":"1.6.0"}
````

@amolnater-qasource can you try this again with debug logging enabled and upload the diagnostics? This is a real bug I just want as much detail about it as possible.

**Comment by cmacknz** [comment]
I think this is an easy way to reproduce that the agent gets stuck in the Updating state if the upgrade fails for any reason. Using an invalid URL for the artifact is just a very convenient way to trigger this problem.

**Comment by amolnater-qasource** [comment]
Hi @cmacknz 

Thank you for looking into this.
We have revalidated this issue on latest 8.7.0 BC3 and triggered upgrade after setting logging level debug for the agent.

**Screenshot:**
![image](https://user-images.githubusercontent.com/77374876/221804467-ecb138de-5d9e-45d5-ae74-5d156e74373c.png)

**Agent Logs:**
[elastic-agent-diagnostics-2023-02-28T07-22-23Z-00.zip](https://github.com/elastic/elastic-agent/files/10848353/elastic-agent-diagnostics-2023-02-28T07-22-23Z-00.zip)

Please let us know if anything else is required from our end.
Thanks!

**Comment by juliaElastic** [comment]
@cmacknz Could you help investigate this? I looked at the logs, and I see the error with the artifacts url, but there is nothing related to upgrade after that. The debug logs have a lot of noise though.

```
{"log.level":"info","@timestamp":"2023-02-28T07:18:14.565Z","log.origin":{"file.name":"upgrade/upgrade.go","file.line":116},"message":"Upgrading agent","log":{"source":"elastic-agent"},"version":"8.7.0","source_uri":"","ecs.version":"1.6.0"}
{"log.level":"warn","@timestamp":"2023-02-28T07:18:14.571Z","log.logger":"transport","log.origin":{"file.name":"transport/tcp.go","file.line":52},"message":"DNS lookup failure \"test-artifacts.elastic.co\": lookup test-artifacts.elastic.co: no such host","ecs.version":"1.6.0"}
```

**Comment by cmacknz** [comment]
Yes I'll take another look, it is possible the agent isn't acknowledging the upgrade here.

**Comment by cmacknz** [comment]
I definitely think there are ways that the upgrade acknowledgment can get lost. The situation in this issue is almost the best case though where we can detect the upgrade will never complete as soon as we attempt to download the upgrade artifact. 

We should be acknowledging the action in this case, so I am inclined to think the problem is on the agent side of things here.

**Comment by cmacknz** [comment]
Well, we aren't automatically setting the Error field of an action acknowledgement for non-retryable errors (see the one call to `SetError` for retryable errors):

https://github.com/elastic/elastic-agent/blob/e1b4c211cb59ca62308ee7b67e5fd351a037bd52/internal/pkg/agent/application/dispatcher/dispatcher.go#L135-L154

The upgrade action is a retryable error however, as it satisfies the retryable action interface:

<details>

https://github.com/elastic/elastic-agent/blob/e1b4c211cb59ca62308ee7b67e5fd351a037bd52/internal/pkg/fleetapi/action.go#L64-L84

</details>

https://github.com/elastic/elastic-agent/blob/e1b4c211cb59ca62308ee7b67e5fd351a037bd52/internal/pkg/fleetapi/action.go#L294-L295

So we should be acking it here:

https://github.com/elastic/elastic-agent/blob/e1b4c211cb59ca62308ee7b67e5fd351a037bd52/internal/pkg/agent/application/dispatcher/dispatcher.go#L247-L261
